subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java

```diff
@@ -229,6 +230,15 @@ private ExcludeSpec doUnion(Set<ExcludeSpec> specs) {
                 elements = Arrays.stream(asArray).filter(Objects::nonNull).collect(toSet());
             }
         }
+        if (elements.size() == 2) {
+            // Corner case to handle one of the two elements being an anyOf
+            Iterator<ExcludeSpec> specIterator = elements.iterator();
+            ExcludeSpec first = specIterator.next();
+            ExcludeSpec second = specIterator.next();
+            if (first instanceof ExcludeAnyOf || second instanceof ExcludeAnyOf) {
+                return anyOf(first, second);
```

Comment:
I wonder if we need a safe exit condition if the collection is the same as the input or we might loop.

Reply:
Damn, you are right. Went too fast on that one.
I should not delegate, I should flatten here and there.

Reply:
Fixed